[RyuJIT/ARM32] Enable GT_MUL_LONG codegen

mskvortsov opened the pull request:
Enables codegen for the int32×int32→int64 multiplication, e.g. for the following method
```csharp
public static long LongMul(int x, int y) { return (long)x * y; }
```
the JIT now generates
```
000008  FB80 1001      smull   r1, r0, r0, r1
00000C  9100           str     r1, [sp]
00000E  9001           str     r0, [sp+0x04]
000010  9800           ldr     r0, [sp]
000012  9901           ldr     r1, [sp+0x04]
```

cc @dotnet/arm32-contrib @dotnet/jit-contrib

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -1581,9 +1581,17 @@ GenTree* DecomposeLongs::DecomposeMul(LIR::Use& use)
     Range().Remove(op2);
 
+#if defined(_TARGET_X86_)
     tree->gtOp.gtOp1 = op1->gtGetOp1();
```
Should update the header comment, since it says"edx:eax" -- say something about arm?

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -1581,9 +1581,17 @@ GenTree* DecomposeLongs::DecomposeMul(LIR::Use& use)
+    use.ReplaceWith(m_compiler, mulTree);
+    Range().Remove(tree);
+#endif
 
```
I think you can avoid this `#ifdef` entirely by changing `SetOperRaw` to `SetOper`. I'm not sure x86 depends on the `Raw` version, or that just happens to work.

BruceForstall commented on src/jit/codegenarmarch.cpp:
```diff
@@ -189,6 +189,21 @@ void CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
+            genProduceReg(treeNode);
+        }
+        break;
+#endif // _TARGET_ARM_
```
Please create a `genCodeForMulLong` function to do this.

BruceForstall commented on src/jit/gtlist.h:
```diff
@@ -186,8 +186,8 @@ GTNODE(SUB_LO           , "-Lo"          ,GenTreeOp          ,0,GTK_BINOP)
-GTNODE(MUL_LONG         , "*long"        ,GenTreeOp          ,1,GTK_BINOP) // A mul that returns the 2N bit result of an NxN multiply. This op
-                                                                           // is used for x86 multiplies that take two ints and return a long
+GTNODE(MUL_LONG         , "*long"        ,GenTreeMulLong     ,1,GTK_BINOP) // A mul that returns the 2N bit result of an NxN multiply. This op
+                                                                           // is used for multiplies that take two ints and return a long
```
x86 doesn't need this new struct. I think you should put all the GenTreeMulLong stuff (here, gtstructs.h, gentree.h, etc.) under `#ifdef _TARGET_ARM_`.

BruceForstall commented:
@CarolEidt this change adds GT_MUL_LONG as a special case node that requires two dst registers. The changes to LSRA look simple; are they sufficient? Do you agree with this approach?

CarolEidt commented on src/jit/lsraarm.cpp:
```diff
@@ -741,7 +746,7 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
 
     // We need to be sure that we've set info->srcCount and info->dstCount appropriately
-    assert((info->dstCount < 2) || tree->IsMultiRegCall());
+    assert((info->dstCount < 2) || tree->IsMultiRegCall() || (tree->OperGet() == GT_MUL_LONG));
```
I suspect we may add other nodes in future that return multiple registers. In any case, for clarity I think it would be useful to have a method on `GenTree()` that indicates whether it produces multiple registers, e.g. `IsMultiRegNode()`.

CarolEidt commented:
> The changes to LSRA look simple; are they sufficient? Do you agree with this approach?

Yes, I think the changes are sufficient. I would like to see an abstraction (as mentioned in a comment) on `GenTree` for nodes that produce multiple registers, but I think it's the case that the multi-reg call work pretty much added the necessary support to LSRA.

---
In reply to: [300890347](https://github.com/dotnet/coreclr/pull/11526#issuecomment-300890347) [](ancestors = 300890347)

CarolEidt commented:
:clock1:

BruceForstall commented:
@mmitche I see:
```
Disk has run out of space and cannot get latest changes. Please contact netciadmin@microsoft.com. IOException: No space left on device
```
https://ci.dot.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_flow_prtest/6772/

BruceForstall commented:
@dotnet-bot test Tizen armel Cross Release Build
@dotnet-bot test Ubuntu x64 Checked Build and Test

BruceForstall commented:
@mskvortsov Is this still [WIP], or do you want me to merge it if the tests all pass?

mskvortsov commented:
@BruceForstall Yes, let it be [WIP] for now, because I want to at least squash the commits before merging, but I'm already OOTO today. I'll finish this PR on Monday.

mskvortsov commented:
@BruceForstall I've addressed both your suggestions about SetOperRaw() and IsMultiRegNode(), rebased onto master to test the x86 change, and squashed. Please merge this.

BruceForstall commented:
@dotnet-bot test this please
